source/opt/folding_rules.cpp

```diff
@@ -124,6 +124,102 @@ Instruction* NonConstInput(IRContext* context, const analysis::Constant* c,
       inst->GetSingleWordInOperand(in_op));
 }
 
+std::vector<uint32_t> ExtractInts(uint64_t val) {
+  std::vector<uint32_t> words;
+  words.push_back(static_cast<uint32_t>(val));
+  words.push_back(static_cast<uint32_t>(val >> 32));
+  return words;
+}
+
+std::vector<uint32_t> GetWordsFromNumericScalarOrVectorConstant(
+    analysis::ConstantManager* const_mgr, const analysis::Constant* c) {
+  std::vector<uint32_t> words;
+  if (const auto* float_constant = c->AsFloatConstant()) {
+    uint32_t width = float_constant->type()->AsFloat()->width();
+    assert(width == 32 || width == 64);
+    if (width == 64) {
+      utils::FloatProxy<double> result(float_constant->GetDouble());
+      words = result.GetWords();
+    } else {
+      utils::FloatProxy<float> result(float_constant->GetFloat());
+      words = result.GetWords();
+    }
+  } else if (const auto* int_constant = c->AsIntConstant()) {
+    uint32_t width = int_constant->type()->AsInteger()->width();
+    assert(width == 32 || width == 64);
+    if (width == 64) {
+      uint64_t uval = static_cast<uint64_t>(int_constant->GetU64());
+      words = ExtractInts(uval);
+    } else {
+      words.push_back(static_cast<uint32_t>(int_constant->GetU32()));
+    }
+  } else if (const auto* vec_constant = c->AsVectorConstant()) {
+    if (const auto* vec_type = vec_constant->type()->AsVector()) {
+      if (vec_type->element_type()->AsInteger() == nullptr &&
+          vec_type->element_type()->AsFloat() == nullptr) {
+        return words;
+      }
+    } else {
+      return words;
+    }
+
+    for (const auto* comp : vec_constant->GetComponents()) {
+      auto comp_in_words =
+          GetWordsFromNumericScalarOrVectorConstant(const_mgr, comp);
+      words.insert(words.end(), comp_in_words.begin(), comp_in_words.end());
+    }
+  }
+  return words;
+}
+
+const analysis::Constant* ConvertWordsToNumericScalarOrVectorConstant(
+    analysis::ConstantManager* const_mgr, const std::vector<uint32_t>& words,
+    const analysis::Type* type) {
+  if (type->AsInteger() || type->AsFloat())
+    return const_mgr->GetConstant(type, words);
+
+  if (const auto* vec_type = type->AsVector()) {
+    const auto* element_type = vec_type->element_type();
+    uint32_t width_in_bits = 0;
+    if (const auto* float_type = element_type->AsFloat())
+      width_in_bits = float_type->width();
+    else if (const auto* int_type = element_type->AsInteger())
+      width_in_bits = int_type->width();
```

Comment:
Getting the width in bit of the element type of a vector could be moved to the vector type class, if it does not already exist.  That seems like code that could be easily reused.

Reply:
done